[tuya] HANDLER_INITIALIZING_ERROR A unit hint must not be set if the item type is not a number with dimension!

I added a new device to my network. It is as tuya Wifi switch with measuring capabilities. The device works fine in the SmartLifew app but not in my openHAB openHAB 5.1.3.
After discovering the newly added device gives an error message:
`Status: 
UNINITIALIZED
HANDLER_INITIALIZING_ERROR
A unit hint must not be set if the item type is not a number with dimension!`

Normally it obvious gets online. 

It look there goes something wrong in the discovery. Can I do something to get it to work?

Or can this be solved in the binding. Is anything from my side needed, please tell me.

This is what I get when the Tuya log is set to debug (not more than when no debug log is on):
`4:57:00.361 Log set Tuya=DEBUG (was ERROR)
14:57:09.372 An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler@3ee6856a': A unit hint must not be set if the item type is not a number with dimension!
java.lang.IllegalArgumentException: A unit hint must not be set if the item type is not a number with dimension!
                                                                                                                        at org.openhab.core.thing.type.ChannelType.<init>(ChannelType.java:88) ~[?:?]
                                                                                                                                                                                                        at org.openhab.core.thing.internal.type.StateChannelTypeBuilderImpl$StateChannelTypeImpl.<init>(StateChannelTypeBuilderImpl.java:46) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                        at org.openhab.core.thing.internal.type.StateChannelTypeBuilderImpl.build(StateChannelTypeBu



                                                                                                                                                                                                                                                                                                                                                                                                                                                   uilderImpl.java:93) ~[?:?]
                                at org.openhab.binding.tuya.internal.TuyaChannelTypeProvider.channelTypeFromSchema(TuyaChannelTypeProvider.java:372) ~[?:?]
                                                                                                                                                                at org.openhab.binding.tuya.internal.TuyaChannelTypeProvider.getChannelType(TuyaChannelTypeProvider.java:170) ~[?:?]
                                                                                                                                                                                                                                                                                        at org.openhab.core.thing.type.ChannelTypeRegistry.getChannelType(ChannelTypeRegistry.java:82) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                                at org.openhab.core.thing.type.ChannelTypeRegistry.getChannelType(ChannelTypeRegistry.java:68) ~[?:?]
                                                        at org.openhab.core.thing.internal.ThingHandlerCallbackImpl.createChannelBuilder(ThingHandlerCallbackImpl.java:220) ~[?:?]
                                                                                                                                                                                        at org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler.lambda$10(TuyaDeviceHandler.java:671) ~[?:?]
                                                                                                                                                                                                                                                                                                        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                                at java.util.Iterator.forEachRemaining(Iterator.java:133) ~[?:?]
                at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939) ~[?:?]
                                                                                                                at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
                                                                                                                                                                                                at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
                                                                                                                                                                                                                                                                                        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
                at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[?:?]
                                                                                                        at org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler.addChannels(TuyaDeviceHandler.java:675) ~[?:?]
                                                                                                                                                                                                                        at org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler.initialize(TuyaDeviceHandler.java:579) ~[?:?]
                                                                                                                                                                                                                                                                                                                                        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]       
                                                                                                                                                                                                                                                                                                                                                                                                                                                at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
                                                        at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:149) ~[?:?]
                                                                                                                                                                                at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) ~[?:?]
                                                                                                                                                                                                                                                                at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
                                                                                                                                                                                                                                                                                                                                        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                                                                        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
                                                                                at java.lang.Thread.run(Thread.java:1583) [?:?]
                                                                                                                               14:57:09.384 Exception occurred while initializing handler of thing 'tuya:tuyaDevice:bf0acfb055d6b19164by7k': A unit hint must not be set if the item type is not a number with dimension!
java.lang.IllegalArgumentException: A unit hint must not be set if the item type is not a number with dimension!
                                                                                                                        at org.openhab.core.thing.type.ChannelType.<init>(ChannelType.java:88) ~[?:?]
                                                                                                                                                                                                        at org.openhab.core.thing.internal.type.StateChannelTypeBuilderImpl$StateChannelTypeImpl.<init>(StateChannelTypeBuilderImpl.java:46) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                        at org.openhab.core.thing.internal.type.StateChannelTypeBuilderImpl.build(StateChannelTypeBuilderImpl.java:93) ~[?:?]
                                at org.openhab.binding.tuya.internal.TuyaChannelTypeProvider.channelTypeFromSchema(TuyaChannelTypeProvider.java:372) ~[?:?]
                                                                                                                                                                at org.openhab.binding.tuya.internal.TuyaChannelTypeProvider.getChannelType(TuyaChannelTypeProvider.java:170) ~[?:?]
                                                                                                                                                                                                                                                                                        at org.openhab.core.thing.type.ChannelTypeRegistry.getChannelType(ChannelTypeRegistry.java:82) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                                at org.openhab.core.thing.type.ChannelTypeRegistry.getChannelType(ChannelTypeRegistry.java:68) ~[?:?]
                                                        at org.openhab.core.thing.internal.ThingHandlerCallbackImpl.createChannelBuilder(ThingHandlerCallbackImpl.java:220) ~[?:?]
                                                                                                                                                                                        at org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler.lambda$10(TuyaDeviceHandler.java:671) ~[?:?]
                                                                                                                                                                                                                                                                                                        at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                                at java.util.Iterator.forEachRemaining(Iterator.java:133) ~[?:?]
                at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939) ~[?:?]
                                                                                                                at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
                                                                                                                                                                                                at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
                                                                                                                                                                                                                                                                                        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
                at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682) ~[?:?]
                                                                                                        at org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler.addChannels(TuyaDeviceHandler.java:675) ~[?:?]
                                                                                                                                                                                                                        at org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler.initialize(TuyaDeviceHandler.java:579) ~[?:?]
                                                                                                                                                                                                                                                                                                                                        at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]       
                                                                                                                                                                                                                                                                                                                                                                                                                                                at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
                                                        at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:149) ~[?:?]
                                                                                                                                                                                at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) ~[?:?]
                                                                                                                                                                                                                                                                at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
                                                                                                                                                                                                                                                                                                                                        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
                                                                                                                                                                                                                                                                                                                                                                                                                                        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
                                                                                at java.lang.Thread.run(Thread.java:1583) [?:?]
                                                                                                                               14:57:23.510 Log set Tuya=ERROR (was DEBUG)`

I have the exact same problem and was just about to open an issue about that.

I am trying [this device](https://pt.aliexpress.com/item/1005004314522092.html) which connects perfectly fine with Tuya and was easily discovered in OH Tuya Binding.

Problem is, the Thing added now reports:

HANDLER_INITIALIZING_ERROR
A unit hint must not be set if the item type is not a number with dimension!

I’ve even tried to add it manually, by adding a Generic Tuya device and inserting all necessary data, but unfortunately it gives me the exact message.

While using the Tuya IoT Platform, I can see for this device this information regarding the Standard Status Set:

```
temp_current	Integer	
{
  "unit": "℃",
  "min": -550,
  "max": 1250,
  "scale": 1,
  "step": 1
}
humidity_value	Integer	
{
  "unit": "%",
  "min": 0,
  "max": 1000,
  "scale": 1,
  "step": 1
}
battery_state	Enum	
{
  "range": [
    "low",
    "middle",
    "high"
  ]
}
charge_state	Boolean	
"{true,false}"
temp_unit_convert	Enum	
{
  "range": [
    "c",
    "f"
  ]
}
maxtemp_set	Integer	
{
  "unit": "℃",
  "min": -550,
  "max": 1250,
  "scale": 1,
  "step": 1
}
minitemp_set	Integer	
{
  "unit": "℃",
  "min": -550,
  "max": 1250,
  "scale": 1,
  "step": 1
}
maxhum_set	Integer	
{
  "unit": "%",
  "min": 0,
  "max": 1000,
  "scale": 1,
  "step": 1
}
minihum_set	Integer	
{
  "unit": "%",
  "min": 0,
  "max": 1000,
  "scale": 1,
  "step": 1
}
temp_alarm	Enum	
{
  "range": [
    "loweralarm",
    "upperalarm",
    "cancel"
  ]
}
hum_alarm	Enum	
{
  "range": [
    "loweralarm",
    "upperalarm",
    "cancel"
  ]
}
bright_value	Integer	
{
  "unit": "%",
  "min": 0,
  "max": 100,
  "scale": 0,
  "step": 1
}
temp_periodic_report	Integer	
{
  "unit": "秒",
  "min": 30,
  "max": 3600,
  "scale": 0,
  "step": 1
}
switch	Boolean	
"{true,false}"
temp_calibration	Integer	
{
  "unit": "℃",
  "min": -30,
  "max": 30,
  "scale": 1,
  "step": 1
}
hum_calibration	Integer	
{
  "unit": "%",
  "min": -200,
  "max": 200,
  "scale": 1,
  "step": 1
}
alarm_time	Integer	
{
  "unit": "秒",
  "min": 1,
  "max": 3600,
  "scale": 0,
  "step": 1
}
beep	Boolean	
"{true,false}"

```
I can also see the Standard Instruction Set:

```
temp_unit_convert	Enum	
{
  "range": [
    "c",
    "f"
  ]
}
maxtemp_set	Integer	
{
  "unit": "℃",
  "min": -550,
  "max": 1250,
  "scale": 1,
  "step": 1
}
minitemp_set	Integer	
{
  "unit": "℃",
  "min": -550,
  "max": 1250,
  "scale": 1,
  "step": 1
}
maxhum_set	Integer	
{
  "unit": "%",
  "min": 0,
  "max": 1000,
  "scale": 1,
  "step": 1
}
minihum_set	Integer	
{
  "unit": "%",
  "min": 0,
  "max": 1000,
  "scale": 1,
  "step": 1
}
temp_periodic_report	Integer	
{
  "unit": "秒",
  "min": 30,
  "max": 3600,
  "scale": 0,
  "step": 1
}
switch	Boolean	
"{true,false}"
temp_calibration	Integer	
{
  "unit": "℃",
  "min": -30,
  "max": 30,
  "scale": 1,
  "step": 1
}
hum_calibration	Integer	
{
  "unit": "%",
  "min": -200,
  "max": 200,
  "scale": 1,
  "step": 1
}
alarm_time	Integer	
{
  "unit": "秒",
  "min": 1,
  "max": 3600,
  "scale": 0,
  "step": 1
}
beep	Boolean	
"{true,false}"

```
In the JSON editor, I have these options:

```
[
  {
    "code": "temp_unit_convert",
    "value": "c"
  },
  {
    "code": "maxtemp_set",
    "value": 1250
  },
  {
    "code": "minitemp_set",
    "value": -400
  },
  {
    "code": "maxhum_set",
    "value": 1000
  },
  {
    "code": "minihum_set",
    "value": 0
  },
  {
    "code": "temp_periodic_report",
    "value": 60
  },
  {
    "code": "switch",
    "value": true
  },
  {
    "code": "temp_calibration",
    "value": 0
  },
  {
    "code": "hum_calibration",
    "value": 0
  },
  {
    "code": "alarm_time",
    "value": 15
  },
  {
    "code": "beep",
    "value": true
  }
]
```

Product name appears like so: (停产)WIFI温湿度报警器

This is what I see in my logs:

```
15:59:48.069[INFO] [openhab.event.ThingStatusInfoEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' updated: UNINITIALIZED
--
15:59:48.074[INFO] [openhab.event.ThingStatusInfoEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' updated: UNINITIALIZED (DISABLED)
15:59:49.283[INFO] [openhab.event.ThingStatusInfoEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' updated: INITIALIZING
15:59:49.318[ERROR] [org.openhab.core.internal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler@6678f42a': A unit hint must not be set if the item type is not a number with dimension!
15:59:49.323[INFO] [openhab.event.ThingStatusInfoEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' updated: UNINITIALIZED (HANDLER_INITIALIZING_ERROR): A unit hint must not be set if the item type is not a number with dimension!
15:59:49.327[ERROR] [org.openhab.core.thing.internal.ThingManagerImpl] - Exception occurred while initializing handler of thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh': A unit hint must not be set if the item type is not a number with dimension!
15:59:50.146[INFO] [openhab.event.ThingStatusInfoEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' updated: UNINITIALIZED (HANDLER_MISSING_ERROR)
15:59:50.149[INFO] [openhab.event.ThingStatusInfoEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' updated: INITIALIZING
15:59:50.180[ERROR] [org.openhab.core.internal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.initialize()' on 'org.openhab.binding.tuya.internal.handler.TuyaDeviceHandler@d277be8': A unit hint must not be set if the item type is not a number with dimension!
15:59:50.181[INFO] [openhab.event.ThingStatusInfoEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' updated: UNINITIALIZED (HANDLER_INITIALIZING_ERROR): A unit hint must not be set if the item type is not a number with dimension!
15:59:50.190[ERROR] [org.openhab.core.thing.internal.ThingManagerImpl] - Exception occurred while initializing handler of thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh': A unit hint must not be set if the item type is not a number with dimension!
15:59:50.192[INFO] [openhab.event.ThingUpdatedEvent] - Thing 'tuya:tuyaDevice:bf26d2530f8c054a6crhvh' has been updated.
```

If any other information might be needed in order to fix the binding, please let me know so I can pass that and make it work properly.

Thank you very much!

--- a/bundles/org.openhab.binding.tuya/src/main/java/org/openhab/binding/tuya/internal/TuyaChannelTypeProvider.java
+++ b/bundles/org.openhab.binding.tuya/src/main/java/org/openhab/binding/tuya/internal/TuyaChannelTypeProvider.java
@@ -370,7 +370,7 @@ public class TuyaChannelTypeProvider implements ChannelTypeProvider {
         } catch (URISyntaxException e) {
         }
 
-        if (!schemaDp.unit.isEmpty()) {
+        if (!schemaDp.unit.isEmpty() && acceptedItemType.startsWith(NUMBER + ":")) {
             channelTypeBuilder.withUnitHint(schemaDp.unit);
         }
 

--- a/bundles/org.openhab.binding.tuya/src/test/java/org/openhab/binding/tuya/internal/TuyaChannelTypeProviderTest.java
+++ b/bundles/org.openhab.binding.tuya/src/test/java/org/openhab/binding/tuya/internal/TuyaChannelTypeProviderTest.java
@@ -1,0 +1,136 @@
+/*
+ * Copyright (c) 2010-2026 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.tuya.internal;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.lenient;
+import static org.openhab.binding.tuya.internal.TuyaBindingConstants.BINDING_ID;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.openhab.binding.tuya.internal.util.SchemaDp;
+import org.openhab.core.thing.i18n.ChannelTypeI18nLocalizationService;
+import org.openhab.core.thing.type.ChannelType;
+import org.openhab.core.thing.type.ChannelTypeUID;
+
+/**
+ * The {@link TuyaChannelTypeProviderTest} verifies that the channel type provider
+ * only sets a unit hint when the accepted item type is a {@code Number:<Dimension>}.
+ * <p>
+ * Previously, an unparseable or dimensionless unit on a {@code value}-type schema
+ * produced an item type of {@code Number} while still calling
+ * {@code withUnitHint(...)}, which made {@code ChannelType} throw
+ * {@code IllegalArgumentException: A unit hint must not be set if the item type is
+ * not a number with dimension!} and left the thing in
+ * {@code HANDLER_INITIALIZING_ERROR}. See issue #20616.
+ *
+ * @author Caglar Eker - Initial contribution
+ */
+@ExtendWith(MockitoExtension.class)
+@NonNullByDefault
+public class TuyaChannelTypeProviderTest {
+
+    private @Mock @NonNullByDefault({}) ChannelTypeI18nLocalizationService localizationServiceMock;
+
+    private static final String PRODUCT_ID = "testprod";
+
+    @BeforeEach
+    public void setUp() {
+        // Return the input channel type unchanged so the test focuses on
+        // TuyaChannelTypeProvider behavior, not on localization.
+        lenient().when(localizationServiceMock.createLocalizedChannelType(any(), any(ChannelType.class), any()))
+                .thenAnswer(inv -> inv.getArgument(1));
+        TuyaSchemaDB.cache.put(PRODUCT_ID, new ConcurrentHashMap<>());
+    }
+
+    @AfterEach
+    public void tearDown() {
+        TuyaSchemaDB.cache.remove(PRODUCT_ID);
+    }
+
+    @Test
+    public void parseableDimensionalUnitSetsNumberDimensionAndUnitHint() {
+        SchemaDp dp = valueDp("°C");
+        putDp("temp", dp);
+
+        ChannelType ct = getChannelType("temp");
+
+        assertNotNull(ct);
+        assertEquals("Number:Temperature", ct.getItemType());
+        assertEquals("°C", ct.getUnitHint());
+    }
+
+    @Test
+    public void unparseableUnitKeepsPlainNumberAndOmitsUnitHint() {
+        // A unit string openHAB can't parse used to trigger
+        // "A unit hint must not be set if the item type is not a number with dimension!"
+        SchemaDp dp = valueDp("bogusunit");
+        putDp("garbage", dp);
+
+        ChannelType ct = getChannelType("garbage");
+
+        assertNotNull(ct);
+        assertEquals("Number", ct.getItemType());
+        assertNull(ct.getUnitHint(), "unit hint must not be set on plain Number item type");
+    }
+
+    @Test
+    public void boolSchemaProducesSwitchWithoutUnitHint() {
+        // Defensive check: even if unit is somehow populated, non-numeric item
+        // types must never receive a unit hint.
+        SchemaDp dp = new SchemaDp();
+        dp.type = "bool";
+        dp.label = "Power";
+        dp.unit = "";
+        putDp("power", dp);
+
+        ChannelType ct = getChannelType("power");
+
+        assertNotNull(ct);
+        assertEquals("Switch", ct.getItemType());
+        assertNull(ct.getUnitHint());
+    }
+
+    private SchemaDp valueDp(String unit) {
+        SchemaDp dp = new SchemaDp();
+        dp.type = "value";
+        dp.label = "Test";
+        dp.unit = unit;
+        return dp;
+    }
+
+    private void putDp(String channelTypeId, SchemaDp dp) {
+        Map<String, SchemaDp> schema = TuyaSchemaDB.cache.get(PRODUCT_ID);
+        assertNotNull(schema);
+        schema.put(channelTypeId, dp);
+    }
+
+    private ChannelType getChannelType(String channelTypeId) {
+        TuyaChannelTypeProvider provider = new TuyaChannelTypeProvider(localizationServiceMock);
+        ChannelTypeUID uid = new ChannelTypeUID(BINDING_ID, PRODUCT_ID + "_" + channelTypeId);
+        ChannelType ct = provider.getChannelType(uid, null);
+        assertNotNull(ct);
+        return ct;
+    }
+}

